src/lib/db.c

```diff
@@ -1823,6 +1823,236 @@ static CK_RV dbup_handler_from_4_to_5(sqlite3 *updb) {
     return rv;
 }
 
+#define CKR_VENDOR_SKIP (CKR_VENDOR_DEFINED | 0x01)
+
+static CK_RV handle_ECDSA_5_to_6(tobject *tobj) {
+
+    CK_OBJECT_CLASS cka_class = attr_list_get_CKA_CLASS(tobj->attrs, CK_OBJECT_CLASS_BAD);
+    if (cka_class != CKO_PRIVATE_KEY) {
+        return CKR_VENDOR_SKIP;
+    }
+
+    CK_KEY_TYPE cka_key_type = attr_list_get_CKA_KEY_TYPE(tobj->attrs, CKA_KEY_TYPE_BAD);
+    if (cka_key_type != CKK_EC) {
+        return CKR_VENDOR_SKIP;
+    }
+
+    CK_ATTRIBUTE_PTR a = attr_get_attribute_by_type(tobj->attrs, CKA_ALLOWED_MECHANISMS);
+
+    CK_ULONG mech_num = a ? (a->ulValueLen/sizeof(CK_MECHANISM_TYPE)) : 0;
+    /* plus three: SHA256 SHA384 and SHA512 */
+    safe_adde(mech_num, 3);
+
+    CK_ULONG total_bytes = 0;
+    safe_mul(total_bytes, mech_num, sizeof(CK_MECHANISM_TYPE));
+
+    CK_MECHANISM_TYPE *new_mechs = calloc(mech_num, sizeof(CK_MECHANISM_TYPE));
+    if (!new_mechs) {
+        tobject_free(tobj);
+        return CKR_HOST_MEMORY;
+    }
+
+    /* copy the old mechanism into the list and add CBC_PAD and CTR */
+    CK_ULONG i;
+    for (i=0; i < mech_num - 1; i++) {
+        assert(a && a->pValue && a->ulValueLen);
+        CK_MECHANISM_TYPE old_mech = ((CK_MECHANISM_TYPE_PTR)(a->pValue))[i];
+        if (old_mech == CKM_ECDSA_SHA256 ||
+                old_mech == CKM_ECDSA_SHA384 ||
+                old_mech == CKM_ECDSA_SHA512) {
+            /*
+             * don't add the ones were adding, just to make the add below
+             * unconditional. This is an unlikely case as someone would have
+             * had to twiddle their object config by hand.
+             */
+            continue;
```

Comment:
This `continue;` while `i` is still incremented seems strange. This means if the condition is true, `new_mechs` will contain an entry with value `0`. Wouldn't it be better if the entry was simply skipped? (For example with new variables `new_i` and `new_mech_num` which would track the state of `new_mechs`)

Reply:
Yep!